ingot-types/src/util.rs

```diff
@@ -345,16 +345,42 @@ where
 /// These are necessary for the type to be used in a variable-length field.
 #[macro_export]
 macro_rules! zerocopy_type {
-    ($t:ty) => {
-        impl $crate::HeaderLen for $t {
+    (
+        $(#[$meta:meta])*
+        $vis:vis struct $name:ident($inner_vis:vis $t:ty)
+    ) => {
```

Comment:
Changing this to just define everything on a tuple struct is very neat, and it works well with the cases we have here. I tend to flip-flop a bit on tuple structs vs `inner: T`, but it's nice here where these elements are 'just' a slice of bytes internally.

The only issue I can think of is that the macro, previously, would also have been applicable to a named struct? I guess that what's here isn't really a downgrade in that regard apart from creating a newtype for the `Emit`/`HeaderLen` impls, since the user would always have needed to copy in the large derive block. If we can also make that work, it'd probaby be useful?

Reply:
Okay, I've refactored the macro to support both tuple and named-field structs, and reverted the changes to `Ipv4Addr` and `Ipv6Addr` (so they're back to having `inner` fields).